ksql-engine/src/main/java/io/confluent/ksql/analyzer/MutableAggregateAnalysis.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.analyzer;
+
+import com.google.common.collect.ImmutableSet;
+import io.confluent.ksql.parser.tree.DereferenceExpression;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.FunctionCall;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class MutableAggregateAnalysis implements AggregateAnalysis {
+
+  private final Set<DereferenceExpression> requiredColumns = new HashSet<>();
```

Comment:
Agreed that we definitely need maintain consistent ordering. As above, I think the order is maintained and I've added tests to prove this. See `shouldUseConsistentOrderInPreAggSelectMapper` in `AggregateNodeTest`, which confirms the order of the pre-select mapper.

Would appreciate your time to double check me here, as I agree its very important and it could just be the test works out the right order, but it's not guaranteed.

Reply:
Looks like you could be right on this one... working on it.

Reply:
Ok, fixed.